### ISSUE TITLE
Hyperparameter Search for AM Datasets + Approaches

### ISSUE THREAD

**Issue by ArneBinder** [ArneBinder/pie-document-level#103]
Up to this point, we have integrated 6 different AM datasets and two model approaches. However, the model performance does not look so well when using the default config parameters. So we need to find better hyperparameters.

Parameter candidates:
 - learning rate
 - transformer base model

What else? Let us extend this list. Maybe we can get some inspiration from: https://github.com/google-research/tuning_playbook

We should have individual issues for all the combinations and link them in the table below.

| datatasets    | pipeline<br/>naive: token classification > RE text classification | joined<br/>generative (BART), see #26 |
|---------------|:----------------------------------------------------------------:|:----------------------------:|
| aae2          |                                                                  |                              |
| abstrct       |                                                                  |                              |
| argmicro      |                                                                  |                              |
| cdcp          |                                                                  |                              |
| sciarg        |                                                                  |                              |
| scidtb_argmin |                                                                  |                              |

**Comment by jk-tripathy** [comment]
@ArneBinder, so I looked into what everyone uses as their sweep/defaults, and I put together this. Another interesting thing to test is whether stacking the seq2seq module changes anything, but I wonder if it is essential to try. It might be interesting to see if having a shorter seq len would allow the model to grasp the data better, but again, I'm not sure about that.

For the batch sizes and the overall number of seeds, I think ill have to play around on the GPUs directly before coming up with something that fits into the experiment + they would probably be different between the different datasets.

## Token Classification

|     Hyperparam    |                                                    Sweep                                                   |
|:-----------------:|:----------------------------------------------------------------------------------------------------------:|
|     batch size    |                                                      -                                                     |
|        seed       |                                                      -                                                     |
|         lr        |                                        5e-5, 3e-5, 2e-5, 3e-4, 1e-4                                        |
|         task lr        |                                        5e-5, 3e-5, 2e-5, 3e-4, 1e-4                                        |
|     base model    | bert-base-uncased, bert-large-uncased, roberta-base, microsoft/deberta-v3-base, microsoft/deberta-v3-large, allenai/longformer-base-4096 |
|  base sci models  |                                      allenai/scibert_scivocab_uncased, allenai/longformer-base-4096, allenai/longformer-scico                                     |
|   max seq length  |                                                128, 256, 512                                               |
| freeze base model |                                                 true, false                                                |
|  seq2seq enc type |                                         linear, lstm, gru, dropout                                         |
|        crf        |                                                 true, false                                                |

**Comment by jk-tripathy** [comment]
For clarification, the taskmodule that I used to align these sweeps is based on the PR found [here](https://github.com/ArneBinder/pie-models/pull/2#issue-1863322341).

**Comment by ArneBinder** [comment]
Cool, thanks for the table! I just added `allenai/longformer-base-4096` to the base (sci) models. And I guess `allenai/longformer-scico` could be useful as base sci model? Maybe there are more useful models that tare pre-trained on scientific text?

**Comment by jk-tripathy** [comment]
oh cool gotcha

**Comment by ArneBinder** [comment]
Regarding the 

> For clarification, the taskmodule that I used to align these sweeps is based on the PR found [here](https://github.com/ArneBinder/pie-models/pull/2#issue-1863322341).

I also just merged https://github.com/ArneBinder/pie-models/pull/3 and https://github.com/ArneBinder/pie-document-level/pull/116, please have a look. The new model implements a parameter `task_learning_rate` that may be interesting: sometimes the classifier head needs a higher learning rate than the base transformer model. I guess it is because the base transformer model is already trained, but not 100% sure. A former colleague mentioned that once. 

So, it would be nice to have the same functionality also in the model-with-seq2seqencoder-and-crf.  Could you add this to [the respective PR](https://github.com/ArneBinder/pie-models/pull/2#issue-1863322341)? I think you can just copy the code from the other PR.

**Comment by jk-tripathy** [comment]
I have also noticed bumping up the clf head LR sometimes makes the model perform better, especially when the base model is not frozen. And cool ill get on that PR

**Comment by ArneBinder** [comment]
Regarding the `seq2seq enc type` in the table: Here it makes often sense to have a combination of them in a stacked fashion. E.g. like in the `sam_paper` config: `sequential(dopout, lstm)`, but we may think about other combinations. This is especially necessary when freeing the base model.

Did you mean that with "Another interesting thing to test is whether stacking the seq2seq module changes anything, but I wonder if it is essential to try."?

**Comment by ArneBinder** [comment]
> I have also noticed bumping up the clf head LR sometimes makes the model perform better, especially when the base model is not frozen. And cool ill get on that PR

What exactly do you mean with "clf head LR"?

**Comment by jk-tripathy** [comment]
No i actually meant stacking the RNN layers themself ie the `num_layer` param in pytorch. Right now if I'm not wrong they are only single layers?

**Comment by jk-tripathy** [comment]
> > I have also noticed bumping up the clf head LR sometimes makes the model perform better, especially when the base model is not frozen. And cool ill get on that PR
> 
> What exactly do you mean with "clf head LR"?

classification head learning rate**

**Comment by jk-tripathy** [comment]
> No i actually meant stacking the RNN layers themself ie the `num_layer` param in pytorch. Right now if I'm not wrong they are only single layers?

I'm not sure if the gradients are the same if you do `sequential(lstm, lstm)` or if you stack them using `num_layers`

**Comment by ArneBinder** [comment]
> No i actually meant stacking the RNN layers themself ie the `num_layer` param in pytorch. Right now if I'm not wrong they are only single layers?

see https://github.com/ArneBinder/pie-document-level/blob/token_classification_model_with_crf/configs/experiment/sciarg_adus_sam_paper.yaml#L69

**Comment by jk-tripathy** [comment]
oh my bad, i must have missed that. thank you

**Comment by ArneBinder** [comment]
> > No i actually meant stacking the RNN layers themself ie the `num_layer` param in pytorch. Right now if I'm not wrong they are only single layers?
> 
> I'm not sure if the gradients are the same if you do `sequential(lstm, lstm)` or if you stack them using `num_layers`

I would do that in a sequential fashion just if you want to do sth in between (e.g. `sequential(lstm, linear, activation, dropout, lstm)`). But let's not overcomplicate it ;)

**Comment by jk-tripathy** [comment]
> > > No i actually meant stacking the RNN layers themself ie the `num_layer` param in pytorch. Right now if I'm not wrong they are only single layers?
> > 
> > 
> > I'm not sure if the gradients are the same if you do `sequential(lstm, lstm)` or if you stack them using `num_layers`
> 
> I would do that in a sequential fashion just if you want to do sth in between (e.g. maybe linear + activation?). But let's not overcomplicate it ;)

got it 😅

**Comment by ArneBinder** [comment]
> > > I have also noticed bumping up the clf head LR sometimes makes the model perform better, especially when the base model is not frozen. And cool ill get on that PR
> > 
> > 
> > What exactly do you mean with "clf head LR"?
> 
> classification head learning rate**

do we already have this parameter for the token classification model?

**Comment by jk-tripathy** [comment]
> > > > I have also noticed bumping up the clf head LR sometimes makes the model perform better, especially when the base model is not frozen. And cool ill get on that PR
> > > 
> > > 
> > > What exactly do you mean with "clf head LR"?
> > 
> > 
> > classification head learning rate**
> 
> do we already have this parameter for the token classification model?

the new params `freeze_base_model` and `task_learning_rate` are not present in the default token classification model found [here](https://github.com/ChristophAlt/pytorch-ie/blob/main/src/pytorch_ie/models/transformer_token_classification.py).
and the `task_learning_rate` is not present in the new sam crf model found [here](https://github.com/ChristophAlt/pytorch-ie/blob/main/src/pytorch_ie/models/transformer_token_classification.py).

**Comment by ArneBinder** [comment]
> the new params freeze_base_model and task_learning_rate are not present in the default token classification model found [here](https://github.com/ChristophAlt/pytorch-ie/blob/main/src/pytorch_ie/models/transformer_token_classification.py).
and the task_learning_rate is not present in the new sam crf model found [here](https://github.com/ChristophAlt/pytorch-ie/blob/main/src/pytorch_ie/models/transformer_token_classification.py).

ok. In general, we should just use the new model (the one with the crf) from now on.

**Comment by jk-tripathy** [comment]
understood

**Comment by ArneBinder** [comment]
was falsely closed because the PR for some hyperparameter search configs was merged (#123)
